### ISSUE TITLE
Documentation for AST node types

### ISSUE THREAD

**Issue by pauliunas** [typescript-eslint/typescript-eslint#2726]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

not applicable

**Expected Result**

not applicable

**Actual Result**

not applicable

**Additional Info**

This is a meta issue about documentation.

I couldn't find any information about the AST node types, especially the ones that are Typescript-specific. For example, if I'm extending a rule and I come across a place where it's dealing with `TSTypeParameterInstantiation` - how do I know what that means? Ideally, it would be great to have a gallery with code snippets (examples) for every available node type. I know that's a great effort, but even starting to add some JSDoc comments with very brief explanations or short snippets to the code would help a lot.

Here's an example of what I'm thinking of:

**Before**
```ts
export interface CallExpression extends CallExpressionBase {
  type: AST_NODE_TYPES.CallExpression;
}
```

**After**
```ts
/**
 * A function/method call.
 * @example function1<Type1>(arg1);
 */
export interface CallExpression extends CallExpressionBase {
  type: AST_NODE_TYPES.CallExpression;
}
```

Of course, that's an easy example, everyone knows what a call expression is. But for more complex types, this would make writing rules significantly easier.

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `4.6.0` |
| `TypeScript`                           | `3.4.3` |
| `node`                                 | `12.19.0` |

**Comment by paulius-valiunas** [comment]
It's nice to finally have online documentation for the properties of each interface, but it's still difficult to map an interface to a code snippet, which was the primary concern of this issue. As in the example I originally wrote in the issue description, [TSTypeParameterInstantiation](https://typescript-eslint.io/packages/ast-spec/generated/#tstypeparameterinstantiation) still doesn't provide any information about how to recognize this AST node in code. If a developer wants to be efficient at working with the AST (e.g. writing ESLint rules), they have to be able to:

1. Upon seeing a code snippet, understand what AST nodes it comprises. This can currently be done with the help of https://astexplorer.net/
2. Upon encountering an AST type, visualize it as a code snippet in their mind.

Both can be done with the help of a debugger, or simply "printf-debugging", but that significantly slows down any development effort until the developer memorizes the bigger part of the AST types after applying them in practice multiple times. Having some reference material to help deal with problem no.1 would be tremendously helpful. As per @bradzacher's suggestion, I'm able to search through this repo and find [a snapshot containing the type I need](https://github.com/typescript-eslint/typescript-eslint/blob/56bf020081815c1a20fedf897021fc19e0ea1023/packages/ast-spec/src/legacy-fixtures/namespaces-and-modules/fixtures/declare-namespace-with-exported-function/snapshots/1-Babel-AST.shot#L75) and then map it to [the corresponding snippet](https://github.com/typescript-eslint/typescript-eslint/blob/56bf020081815c1a20fedf897021fc19e0ea1023/packages/ast-spec/src/legacy-fixtures/namespaces-and-modules/fixtures/declare-namespace-with-exported-function/fixture.ts), but unfortunately that is a time-consuming process and it would be much more helpful to have each interface in the documentation (or at least some of them) link to a minimal example snippet.

**Comment by JoshuaKGoldberg** [comment]
Good point, thanks - filed #9671 as a followup. 👍
